### PR TITLE
[618] Migrate Courses - Updating age range

### PR DESCRIPTION
**Pull request by goodviber**
### Context
The "Courses" section (linked from a page like this https://qa.publish-teacher-training-courses.service.gov.uk/organisations/2GU/2022) needs to be reimplemented in New Publish.
### Changes proposed in this pull request
Reimplement the functionality to update the age range from the basic details tab view
age_range

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by goodviber** [comment]
[Migrate Courses - Course show page - Updating age range](https://trello.com/c/2ncWXMRF/618-migrate-courses-course-show-page-updating-age-range)

File: app/controllers/publish/courses/age_range_controller.rb [hunk 1]
```diff
@@ -3,6 +3,8 @@ module Courses
     class AgeRangeController < PublishController
       include CourseBasicDetailConcern
       decorates_assigned :course
+      before_action :build_recruitment_cycle, only: %i[edit update]
+      before_action :build_provider, only: %i[edit update]
```
Comment (defong) [inline app/controllers/publish/courses/age_range_controller.rb:7]:
```suggestion
```

It should be using https://github.com/DFE-Digital/teacher-training-api/blob/db1ca9b0b3a8c8d999e544cf1028766f1814b47b/app/controllers/publish/publish_controller.rb#L9-L17

File: app/controllers/publish/courses/age_range_controller.rb [hunk 2]
```diff
@@ -87,12 +85,24 @@ def current_step
...
+        )
+      end
+
+      def build_recruitment_cycle
+        @recruitment_cycle = RecruitmentCycle.find_by(
+          year: params[:recruitment_cycle_year],
+        ) || RecruitmentCycle.current_recruitment_cycle
+      end
+
       def build_course
```
Comment (defong) [inline app/controllers/publish/courses/age_range_controller.rb:102]: Can be dropped and as it should be using

https://github.com/DFE-Digital/teacher-training-api/blob/db1ca9b0b3a8c8d999e544cf1028766f1814b47b/app/models/concerns/publish/course_basic_detail_concern.rb#L60-L62

Comment (defong) [inline app/controllers/publish/courses/age_range_controller.rb:102]: Can be dropped and as it should be using

https://github.com/DFE-Digital/teacher-training-api/blob/db1ca9b0b3a8c8d999e544cf1028766f1814b47b/app/models/concerns/publish/course_basic_detail_concern.rb#L60-L62

File: app/policies/course_policy.rb [hunk 3]
```diff
@@ -39,6 +39,7 @@ def send_vacancies_updated_notification?
 
   alias_method :details?, :show?
   alias_method :update?, :show?
+  alias_method :edit?, :show? # is this the correct permission?
```
Comment (defong) [inline app/policies/course_policy.rb:42]: For the moment it is fine as the basic permission available is 
- if you can see it then you can edit it

Comment (defong) [inline app/policies/course_policy.rb:42]: For the moment it is fine as the basic permission available is 
- if you can see it then you can edit it

File: config/locales/en.yml [hunk 4]
```diff
@@ -441,4 +459,4 @@ en:
   date:
     formats:
       default: "%-d %B %Y"
-      short: "%B %Y"  
\ No newline at end of file
+      short: "%B %Y"
```
Comment (defong) [inline config/locales/en.yml:462]: newline please

Comment (defong) [inline config/locales/en.yml:462]: newline please

File: app/controllers/publish/courses/age_range_controller.rb (hunk 2 above)
Comment (goodviber) [reply app/controllers/publish/courses/age_range_controller.rb:102]: Have to call pundit though? 
```
 Pundit::AuthorizationNotPerformedError:
       Publish::Courses::AgeRangeController
```
Something like this?
```
def build_course
      #  @course = @provider.courses.find_by!(course_code: params[:code].upcase)
        super
        authorize @course
      end
```

File: config/locales/en.yml [hunk 5]
```diff
@@ -304,6 +304,24 @@ en:
   activemodel:
     errors:
       models:
+        publish/age_range_form:
+          attributes:
+            age_range_in_years:
+              blank: "You need to pick an age range"
+            course_age_range_in_years_other_from:
+              blank: "Enter an age in From"
+              not_a_number: "Enter a valid age in From"
```
Comment (darokel) [inline config/locales/en.yml:313]:
```suggestion
not_a_number: &valid_message "Enter a valid age in From"
```

Then you could re-use the string via `*valid_message`

not sure if it's worth doing it here but just a little tip for re-usability

Comment (goodviber) [reply config/locales/en.yml:313]: Ta

File: spec/features/publish/courses/editing_age_range_spec.rb [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  scenario "selecting a custom age range" do
+    when_i_visit_the_edit_age_range_page
+    when_i_select_a_custom_age_range
+    and_i_click_continue
+    then_i_should_see_a_success_message
+    and_the_course_age_range_is_updated("10_to_15")
+  end
+
+  scenario "selecting an invalid age range" do
```
Comment (defong) [inline spec/features/publish/courses/editing_age_range_spec.rb:25]:
```suggestion
  scenario "selecting an invalid custom age range" do
```

File: app/controllers/publish/courses/age_range_controller.rb (hunk 2 above)
Comment (darokel) [reply app/controllers/publish/courses/age_range_controller.rb:88]: I have added the permissions in concern directly. eg https://github.com/DFE-Digital/teacher-training-api/pull/2566/files#diff-b6451eb1b3fe8ddfef71d842610dc538ff9a6e6fdb6cbfef6eda1fde91d6b02dR21. Maybe you could do the same and whoevers goes in will take precedence. Think you need to authorise the provider not the course?

File: spec/features/publish/courses/editing_age_range_spec.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  end
+
+private
+
+  def given_i_am_authenticated_as_a_provider_user
+    given_i_am_authenticated(user: create(:user, :with_provider))
+  end
+
+  def and_there_is_a_course_i_want_to_edit
+    given_a_course_exists(enrichments: [build(:course_enrichment, :published)])
```
Comment (defong) [inline spec/features/publish/courses/editing_age_range_spec.rb:40]:
```suggestion
    given_a_course_exists
```

Don't think you need any enrichments

File: app/controllers/publish/courses/age_range_controller.rb (hunk 2 above)
Comment (defong) [reply app/controllers/publish/courses/age_range_controller.rb:88]: I think the `before_action` is not needed as it called via `super`

**Review by defong — APPROVED** [review]

**Pull request merged**